faculty/clients/api.py

```diff
@@ -0,0 +1,635 @@
+
+    @post_load
+    def make_list_apis(self, data, **kwargs):
+        # Flatten the API response
+        return data["apis"]
+
+
+class APIClient(BaseClient):
+    """Client for the Faculty API deployment.
+
+    Either build this client with a session directly, or use the
+    :func:`faculty.client` helper function:
+
+    >>> client = faculty.client("api")
+
+    Parameters
+    ----------
+    url : str
+        The URL of the server management service.
+    session : faculty.session.Session
+        The session to use to make requests.
+    """
+
+    SERVICE_NAME = "aperture"
+
+    def create(
+        self,
+        project_id,
+        api_definition,
+        subdomain,
+        name="",
+        description="",
+        environment_ids=[],
+        default_server_size=None,
+    ):
+        """Create a new API.
+
+        Parameters
+        ----------
+        project_id : uuid.UUID
+            The project to create the API in.
+        api_definition : CommandDefinition
+            The API's command definition: WSGI(Flask)/Plumber/Script(Custom).
+        subdomain : str
+            The subdomain where the API should run.
+        name : str
+            The API's name.
+        description : str
+            The API's description field.
+        environment_ids : List[uuid.UUID]
+            The environments to apply to the API's instances.
+        default_server_size : ServerType
+            The default server size to set
+        """
+        payload = {
+            "definition": CommandDefinitionSchema().dump(api_definition)
+        }
+        payload["subdomain"] = subdomain
+        payload["name"] = name
+        payload["description"] = description
+
+        payload["environmentIds"] = environment_ids
+
+        # TODO: turn these into schema handling as well
+        if default_server_size is None:
+            default_server_size = {
+                "instanceSizeType": "custom",
+                "instanceSize": {"milliCpus": 1000, "memoryMb": 4096},
+            }
+        payload["defaultServerSize"] = default_server_size
+
+        return self._post_raw(
+            "/project/{}/api".format(project_id), json=payload
+        )
+
+    @staticmethod
+    def flask_definition(working_directory, module, wsgi_object):
+        return WSGIDefinition(
+            api_type=APIType.WSGI,
+            working_directory=working_directory,
+            module=module,
+            wsgi_object=wsgi_object,
+        )
+
+    @staticmethod
+    def plumber_definition(working_directory, script_name):
+        return PlumberDefinition(
+            api_type=APIType.PLUMBER,
+            working_directory=working_directory,
+            script_name=script_name,
+        )
+
+    @staticmethod
+    def script_definition(working_directory, script_name):
+        return ScriptDefinition(
+            api_type=APIType.SCRIPT,
+            working_directory=working_directory,
+            script_name=script_name,
+        )
```

Comment:
Is this point of these to aid construction of API definition objects to send to the backend? I don't think these belong as static methods of the client.

Reply:
Removed all of these, people can create the `WSGIDefinition`/`PlumberDefinition`/`ScriptDefinition` objects themselves to pass onto the relevant functions.